### PR TITLE
Grid layout

### PR DESCRIPTION
**Pull request by jaredoconnell**
This PR adds a basic grid layout to Masonry and Xilem.

The way this layout works is it has a fixed grid based on the initial size passed in, and grid items are placed based on the position requested. Grid items are allowed to span more than one cell, if requested.

There are potential improvements that could be done, like the use of intrinsic sizing for varied column width based on content size. Though that could be done in the future taffy layout if we want to keep this one simple.

~~This PR is still a draft because of one remaining concern. I was not able to successfully optimize with conditional calls to child widgets for layout. It led to crashes about the paint rects not being within the widget's paint rect. `Error in 'Grid' #16: paint_rect Rect { x0: 0.0, y0: 0.0, x1: 800.0, y1: 610.0 } doesn't contain paint_rect Rect { x0: 400.5, y0: 0.0, x1: 800.5, y1: 150.0 } of child widget 'Button' #5`. My failed attempt at fixing it is commented out.~~

Since I am rusty on View Sequences, a lot of that code is based on the Flex implementation. Let me know if I did anything incorrectly or if any of it is unnecessary, or if anything is missing.

### CODE REVIEW

File: masonry/examples/grid_masonry.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 the Xilem Authors and the Druid Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This is a very small example of how to setup a masonry application.
+//! It does the almost bare minimum while still being useful.
```
Comment (DJMcNab) [inline masonry/examples/grid_masonry.rs:5]:
```suggestion
//! Shows how to use a grid layout in Masonry.
```

File: masonry/examples/grid_masonry.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+impl DrawnButton {
+    fn get_label(&self) -> String {
+        format!("X: {}, Y: {}, W: {}, H: {}", self.grid_params.x, self.grid_params.y, self.grid_params.width, self.grid_params.height)
+    }
+}
+
+pub fn main() {
+    let label = Label::new("Change spacing by right and\n left clicking on the buttons")
```
Comment (DJMcNab) [inline masonry/examples/grid_masonry.rs:51]:
```suggestion
    let label = Prose::new("Change spacing by right and\n left clicking on the buttons")
```
Labels should generally only be used for text inside other interactive widgets.

File: masonry/examples/grid_masonry.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+struct DrawnButton {
+    grid_params: GridParams,
+}
+
+impl DrawnButton {
+    fn get_label(&self) -> String {
+        format!("X: {}, Y: {}, W: {}, H: {}", self.grid_params.x, self.grid_params.y, self.grid_params.width, self.grid_params.height)
+    }
+}
```
Comment (DJMcNab) [inline masonry/examples/grid_masonry.rs:48]: I don't think this `DrawnButton` abstraction provides much value. Maybe something like this:
```suggestion
fn grid_button(params: GridParams) -> Button {
    Button::new(format!("X: {}, Y: {}, W: {}, H: {}", params.x, params.y, params.width, params.height))
}
```

File: masonry/examples/grid_masonry.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 the Xilem Authors and the Druid Authors
```
Comment (DJMcNab) [inline masonry/examples/grid_masonry.rs:1]:
```suggestion
// Copyright 2024 the Xilem Authors
```

File: masonry/examples/grid_masonry.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        DrawnButton{ grid_params: GridParams { x: 3, y: 1, width: 1, height: 1 }},
+        DrawnButton{ grid_params: GridParams { x: 3, y: 2, width: 1, height: 1 }},
+        DrawnButton{ grid_params: GridParams { x: 0, y: 3, width: 4, height: 1 }},
+    ];
+
+    let driver = Driver {
+        grid_spacing: 1.0,
+    };
+
+    // Arrange the two widgets vertically, with some padding
```
Comment (DJMcNab) [inline masonry/examples/grid_masonry.rs:67]: Is this comment accurate?

File: masonry/examples/grid_masonry.rs
Comment (DJMcNab) [inline masonry/examples/grid_masonry.rs:1]: This example works well. The UX pattern shown is a bit weird, but it achieves the aims.

File: xilem/examples/calc.rs [hunk 6]
```diff
@@ -184,16 +184,20 @@ impl Calculator {
     }
 }
 
+// TODO: Is there a more generic way of doing the return type? This doesn't allow different
+//      "types" of grid_pos values. For example, I couldn't add the operator button.
+fn num_row(nums: [&'static str; 3], row: i32) -> Vec<GridItem<impl WidgetView<Calculator>, Calculator, ()>> {
```
Comment (DJMcNab) [inline xilem/examples/calc.rs:189]: I'd expect this to return `impl GridSequence`

File: xilem/src/view/mod.rs [hunk 7]
```diff
@@ -32,6 +32,9 @@ mod prose;
 pub use prose::*;
 
 mod textbox;
+mod grid;
+pub use grid::*;
```
Comment (DJMcNab) [inline xilem/src/view/mod.rs:36]: I'd probably put this after flex, rather than inside `textbox`.

File: xilem/examples/calc.rs (hunk 6 above)
Comment (jaredoconnell) [reply xilem/examples/calc.rs]: Changed. That works.

File: xilem/src/view/mod.rs (hunk 7 above)
Comment (jaredoconnell) [reply xilem/src/view/mod.rs]: Yeah. That didn't make any sense 😄 Probably a merge mistake or IDE automated task.
Fixed.

File: masonry/examples/grid_masonry.rs
Comment (jaredoconnell) [reply masonry/examples/grid_masonry.rs:1]: Yeah. The way I did it in the example isn't intended for typical use. Probably just spreadsheets and things like calc, where I had the top flex span all four cells.

**Review by jaredoconnell — COMMENTED** [review]
The recent commit addresses most of the issues other than:
- Documentation
- Optimization of layout calls

File: masonry/examples/grid_masonry.rs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    Button::new(format!(
+        "X: {}, Y: {}, W: {}, H: {}",
+        params.x, params.y, params.width, params.height
+    ))
+}
+
+pub fn main() {
+    let label = SizedBox::new(
+        Prose::new("Change spacing by right and\n left clicking on the buttons")
+            .with_text_size(14.0),
```
Comment (PoignardAzur) [inline masonry/examples/grid_masonry.rs:47]: Minor bikeshedding:

```suggestion
        Prose::new("Change spacing by right and\nleft clicking on the buttons")
            .with_text_size(14.0),
```

Also, for what it's worth, text should probably be centered.

File: masonry/src/widget/grid.rs [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    Point, PointerEvent, Size, StatusChange, TextEvent, Widget, WidgetId, WidgetPod,
+};
+
+pub struct Grid {
+    children: Vec<Child>,
+    grid_width: i32,
+    grid_height: i32,
+    grid_spacing: f64,
+    //old_bc: BoxConstraints,
+    //needs_layout: bool,
```
Comment (PoignardAzur) [inline masonry/src/widget/grid.rs:23]: I would probably just remove the layout caching for now. Layout caching is going to be handled directly by Masonry eventually.

File: masonry/src/widget/grid.rs [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+    }
+}
+
+struct Child {
+    widget: WidgetPod<Box<dyn Widget>>,
+    x: i32,
+    y: i32,
+    width: i32,
+    height: i32,
+}
```
Comment (PoignardAzur) [inline masonry/src/widget/grid.rs:264]: I'd rather follow the same order as the rest of the widget files:

- First define the types.
- Then inherent impls for these types.
- Then WidgetMut.
- Then impl Widget.
- Then tests.

Flex doesn't follow that order, but I think it's the only one.

File: masonry/src/widget/grid.rs [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+            grid.remove_child(0);
+        });
+        assert_render_snapshot!(harness, "2x2_with_removed_widget");
+
+        // Add it back
+        harness.edit_root_widget(|mut grid| {
+            let mut grid = grid.downcast::<Grid>();
+            grid.add_child(button::Button::new("A"), GridParams::new(0, 0, 1, 1));
+        });
+        assert_render_snapshot!(harness, "initial_2x2"); // Should be back to the original state
```
Comment (PoignardAzur) [inline masonry/src/widget/grid.rs:402]: Huh. You reuse the same snapshot multiple times. That's pretty clever.

**Review by PoignardAzur — APPROVED** [review]
Overall it's a really well-written PR. I especially appreciate the tests.

There are a few minor things I'd want to change, but the only thing that's truly blocking is the commented-out code. Aside from that, this is good to merge.

File: masonry/examples/grid_masonry.rs (hunk 8 above)
Comment (DJMcNab) [reply masonry/examples/grid_masonry.rs:47]: How about:
```suggestion
        Prose::new("Change spacing by right and\n\
                               left clicking on the buttons")
            .with_text_size(14.0),
```

File: masonry/src/widget/grid.rs (hunk 9 above)
Comment (jaredoconnell) [reply masonry/src/widget/grid.rs]: Removed. I agree that letting masonry handle it is a better way forward.

File: masonry/src/widget/grid.rs (hunk 10 above)
Comment (jaredoconnell) [reply masonry/src/widget/grid.rs]: I changed the order to match what you're requesting. There are several types, so let me know if I didn't handle the ordering of that stuff the way you wanted.

File: masonry/examples/grid_masonry.rs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        "X: {}, Y: {}, W: {}, H: {}",
+        params.x, params.y, params.width, params.height
+    ))
+}
+
+pub fn main() {
+    let label = SizedBox::new(
+        Prose::new(
+            "Change spacing by right and \
+                               left clicking on the buttons",
```
Comment (DJMcNab) [inline masonry/examples/grid_masonry.rs:49]: If we're not going to have a newline here, I'd probably just put it back on one line.
Although I will admit, aligning "left" and "right" here is quite funny.

File: masonry/src/widget/grid.rs (hunk 10 above)
Comment (PoignardAzur) [reply masonry/src/widget/grid.rs]: It's all good. The marker comments are a nice touch.

**Comment by PoignardAzur** [comment]
One more thing I'd like is a test for negative spacing.

Otherwise, no comment, it's a very clean PR.

File: masonry/examples/grid_masonry.rs (hunk 12 above)
Comment (jaredoconnell) [reply masonry/examples/grid_masonry.rs]: Moved back to one line.

**Comment by PoignardAzur** [comment]
Awesome!
